src/vs/workbench/browser/parts/panel/panelPart.ts

```diff
@@ -144,47 +153,162 @@ export class PanelPart extends CompositePart<Panel> implements IPanelService {
 			})
 		}));
 
-		for (const panel of this.getPanels()) {
-			this.compositeBar.addComposite(panel);
-		}
-
 		this.activePanelContextKey = ActivePanelContext.bindTo(contextKeyService);
 		this.panelFocusContextKey = PanelFocusContext.bindTo(contextKeyService);
 
 		this.registerListeners();
+		this.onDidRegisterPanels([...this.getPanels()]);
+	}
+
+	private onDidRegisterPanels(panels: PanelDescriptor[]): void {
+		for (const panel of panels) {
+			const cachedPanel = this.getCachedPanels().filter(({ id }) => id === panel.id)[0];
+			const activePanel = this.getActivePanel();
+			const isActive = activePanel?.getId() === panel.id;
+
+			if (isActive || !this.shouldBeHidden(panel.id, cachedPanel)) {
+				this.compositeBar.addComposite(panel);
+
+				// Pin it by default if it is new
+				if (!cachedPanel) {
+					this.compositeBar.pin(panel.id);
+				}
+
+				if (isActive) {
+					this.compositeBar.activateComposite(panel.id);
+				}
+			}
+		}
+
+		for (const panel of panels) {
+			this.enableCompositeActions(panel);
+			const viewContainer = this.getViewContainer(panel.id);
+			if (viewContainer?.hideIfEmpty) {
+				const viewDescriptors = this.viewsService.getViewDescriptors(viewContainer);
+				if (viewDescriptors) {
+					this.onDidChangeActiveViews(panel, viewDescriptors);
+					this.panelDisposables.set(panel.id, viewDescriptors.onDidChangeActiveViews(() => this.onDidChangeActiveViews(panel, viewDescriptors)));
+				}
+			}
+		}
+	}
+
+	private onDidDeregisterPanel(panelId: string): void {
+		const disposable = this.panelDisposables.get(panelId);
+		if (disposable) {
+			disposable.dispose();
+		}
+
+		this.panelDisposables.delete(panelId);
+		this.hideComposite(panelId);
+	}
+
+	private enableCompositeActions(panel: PanelDescriptor): void {
+		const { activityAction, pinnedAction } = this.getCompositeActions(panel.id);
+		if (activityAction instanceof PlaceHolderPanelActivityAction) {
+			activityAction.setActivity(panel);
+		}
+
+		if (pinnedAction instanceof PlaceHolderToggleCompositePinnedAction) {
+			pinnedAction.setActivity(panel);
+		}
+	}
+
+	private onDidChangeActiveViews(panel: PanelDescriptor, viewDescriptors: IViewDescriptorCollection): void {
+		if (viewDescriptors.activeViewDescriptors.length) {
+			this.compositeBar.addComposite(panel);
+		} else {
+			this.hideComposite(panel.id);
+		}
+	}
+
+	private shouldBeHidden(panelId: string, cachedPanel?: ICachedPanel): boolean {
+		const viewContainer = this.getViewContainer(panelId);
+		if (!viewContainer || !viewContainer.hideIfEmpty) {
+			return false;
+		}
+
+		return cachedPanel?.views && cachedPanel.views.length
+			? cachedPanel.views.every(({ when }) => !!when && !this.contextKeyService.contextMatchesRules(ContextKeyExpr.deserialize(when)))
+			: panelId === TEST_VIEW_CONTAINER_ID /* Hide Test Panel for the first time or it had no views registered before */;
```

Comment:
Checking with `TEST_VIEW_CONTAINER_ID` is specific to Sidebar, so not sure if you need this here

Reply:
originally left this in if we didn't enforce unique ids across location and we chose to re-use the test container, removing.